components/uniset/src/uniset.rs

```diff
@@ -124,25 +125,72 @@ impl UnicodeSet {
         &self.inv_list
     }
 
-    /// Yields an iterator going through the character set in the [`UnicodeSet`]
+    /// Yields an [`Iterator`] going through the character set in the [`UnicodeSet`]
     ///
     /// # Examples
     ///
     /// ```
     /// use icu::uniset::UnicodeSet;
     /// let example_list = vec![65, 68, 69, 70];
     /// let example = UnicodeSet::from_inversion_list(example_list).unwrap();
-    /// let mut example_iter = example.iter_chars();
-    /// assert_eq!(Some('A'), example_iter.next());
-    /// assert_eq!(Some('B'), example_iter.next());
-    /// assert_eq!(Some('C'), example_iter.next());
-    /// assert_eq!(Some('E'), example_iter.next());
-    /// assert_eq!(None, example_iter.next());
+    /// let mut ex_iter_chars = example.iter_chars();
+    /// assert_eq!(Some('A'), ex_iter_chars.next());
+    /// assert_eq!(Some('B'), ex_iter_chars.next());
+    /// assert_eq!(Some('C'), ex_iter_chars.next());
+    /// assert_eq!(Some('E'), ex_iter_chars.next());
+    /// assert_eq!(None, ex_iter_chars.next());
     /// ```
     pub fn iter_chars(&self) -> impl Iterator<Item = char> + '_ {
         self.inv_list.chunks(2).flat_map(|pair| (pair[0]..pair[1])).filter_map(char::from_u32)
     }
 
+    /// Yields an [`Iterator`] returning the ranges of the code points that are
+    /// included in the [`UnicodeSet`]
+    ///
+    /// Ranges are returned as [`RangeInclusive`], which is inclusive of its
+    /// `end` bound value. An end-inclusive behavior matches the ICU4C/J
+    /// behavior of ranges, ex: `UnicodeSet::contains(UChar32 start, UChar32 end)`.
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// use icu::uniset::UnicodeSet;
+    /// let example_list = vec![65, 68, 69, 70];
+    /// let example = UnicodeSet::from_inversion_list(example_list).unwrap();
+    /// let mut example_iter_ranges = example.iter_ranges();
+    /// assert_eq!(Some(65..=67), example_iter_ranges.next());
+    /// assert_eq!(Some(69..=69), example_iter_ranges.next());
+    /// assert_eq!(None, example_iter_ranges.next());
+    /// ```
+    pub fn iter_ranges(&self) -> impl Iterator<Item = RangeInclusive<u32>> + '_ {
+        self.inv_list.chunks(2).map(|pair| RangeInclusive::new(pair[0], pair[1] - 1))
+    }
+
+    /// Returns the number of ranges contained in this [`UnicodeSet`]
+    pub fn get_range_count(&self) -> usize {
+        self.inv_list.len() / 2
+    }
+
+    /// Returns the first code point in the specified range of this [`UnicodeSet`].
+    /// Returns `None` if `idx` is out of bounds.
+    pub fn get_range_start(&self, idx: usize) -> Option<u32> {
```

Comment:
Question: why did you add this (and `get_range_end`)?

Suggestion, Option 1: These seem unnecessary. We have `iter_ranges()`.  I know ICU4C has these methods, but we don't really need it for Rust.  Remember that FFI gets its own set of functions that may be slightly different from the Rust functions.

Suggestion, Option 2: If you have a reason for adding these, please add docs tests.  All public methods should have docs tests by default unless there is a compelling reason not to have them.  (For methods that are closely coupled, it's probably fine to have a docs test on only one of them)

Reply:
Part of the reason is that it is in ICU, and part of the reason is that the performance characteristic is different between `getRangeStart` (and `getRangeEnd`) which is O(1) because it is backed by a vector and `Iterator` which can only be O(n) since there is no random access guarantee. (I am imagining that it is similar to the Java List interface vs. concrete ArrayList in this regard)

Let me know if you think it is unnecessary -- it's certainly easier to remove (option 1). Otherwise, I will add doc tests (option 2).

Reply:
The thing is that `getRangeStart` and `getRangeEnd` only seem useful to me if you call them in a loop, like

```
for i in 0..uniset.get_range_count() {
    let range_start = uniset.get_range_start(i);
    let range_end = uniset.get_range_end(i);
    // ...
}
```

which is `O(n)` overall.  Can you share a use case where these separate methods would be helpful?

Reply:
You're right, I checked with @iainireland , there's no use case that I know of that needs random access on the ranges in the set. So I will remove it, and it's easy enough to add back in if we need it later anyways if we change our minds.

Reply:
One potential complication that I just thought of: it's not clear to me how to safely expose an idiomatic Rust iterator over FFI. How badly can things go wrong if the C++ side mutates the UnicodeSet while there's a live iterator? Is it enough to just include a sternly worded warning in the documentation?

Exposing `get_range_count()` / `get_range_start()` / `get_range_end()` is not idiomatic for Rust, but it might be less error-prone as an FFI interface.

Reply:
FFI/Diplomat has its own set of functions designed for ease of use over FFI. This PR is for the Rust interface, not the FFI interface. When the FFI/Diplomat for UnicodeSet is done, that's the point where we can engage in the interesting conversation about iterators over FFI.